src/drv_oculus_rift/rift.c

```diff
@@ -139,24 +222,213 @@ static void handle_tracker_sensor_msg(rift_priv* priv, unsigned char* buffer, in
+		rift_touch_controller_t *touch, pkt_rift_radio_message *msg)
+{
+	// The top bits are carrying something unknown. Ignore them
+  uint8_t buttons = msg->touch.buttons & 0xf;
+	rift_touch_calibration *c = &touch->calibration;
+
+	if (touch->buttons != buttons) {
+		LOGV ("touch controller %d buttons now %x",
+				touch->base.id, buttons);
+	}
+	touch->buttons = buttons;
+
+	if (!(msg->touch.timestamp ||
+	      msg->touch.accel[0] || msg->touch.accel[1] || msg->touch.accel[2] ||
+	      msg->touch.gyro[0] || msg->touch.gyro[1] || msg->touch.gyro[2]))
+		return;
+
+	if (!touch->have_calibration) {
+		/* We need calibration data to do any more */
+		if (rift_touch_get_calibration (hmd->radio_handle, touch->device_num,
+				&touch->calibration) < 0)
+			return;
+		touch->have_calibration = true;
+	}
+
+	// time in microseconds
+	int32_t dt;
+	if (touch->time_valid)
+		dt = msg->touch.timestamp - touch->last_timestamp;
+	else
+		dt = 0;
+
+	const double dt_s = 1e-6 * dt;
+	double a[3] = {
+		OHMD_GRAVITY_EARTH / 2048 * msg->touch.accel[0],
+		OHMD_GRAVITY_EARTH / 2048 * msg->touch.accel[1],
+		OHMD_GRAVITY_EARTH / 2048 * msg->touch.accel[2],
+	};
+	double g[3] = {
+		2.0 / 2048 * msg->touch.gyro[0],
+		2.0 / 2048 * msg->touch.gyro[1],
+		2.0 / 2048 * msg->touch.gyro[2],
+	};
+	vec3f mag = {{0.0f, 0.0f, 0.0f}};
+	vec3f gyro;
+	vec3f accel;
+
+	/* Apply correction offsets first - bottom row of the
+	 * calibration 3x4 matrix */
+	int i;
+	for (i = 0; i < 3; i++) {
+		a[i] -= c->acc_calibration[9 + i];
+		g[i] -= c->gyro_calibration[9 + i];
+	}
+	/* Then the 3x3 rotation matrix in row-major order */
+	accel.x = c->acc_calibration[0] * a[0] +
+			  c->acc_calibration[1] * a[1] +
+			  c->acc_calibration[2] * a[2];
+	accel.y = c->acc_calibration[3] * a[0] +
+			  c->acc_calibration[4] * a[1] +
+			  c->acc_calibration[5] * a[2];
+	accel.z = c->acc_calibration[6] * a[0] +
+			  c->acc_calibration[7] * a[1] +
+			  c->acc_calibration[8] * a[2];
+	gyro.x = c->gyro_calibration[0] * g[0] +
+			  c->gyro_calibration[1] * g[1] +
+			  c->gyro_calibration[2] * g[2];
+	gyro.y = c->gyro_calibration[3] * g[0] +
+			  c->gyro_calibration[4] * g[1] +
+			  c->gyro_calibration[5] * g[2];
+	gyro.z = c->gyro_calibration[6] * g[0] +
+			  c->gyro_calibration[7] * g[1] +
+			  c->gyro_calibration[8] * g[2];
+
+#if 0
+	printf ("controller %d dt %f a [%f,%f,%f], g[%f,%f,%f]\n",
+	    touch->device_num, dt_s, accel.x, accel.y, accel.z,
+			gyro.x, gyro.y, gyro.z);
+#endif
+	ofusion_update(&touch->imu_fusion, dt_s, &gyro, &accel, &mag);
+	touch->last_timestamp = msg->touch.timestamp;
+	touch->time_valid = true;
+
+	float t;
+	if (msg->touch.trigger < c->trigger_mid_range) {
+		t = 1.0f - ((float)msg->touch.trigger - c->trigger_min_range) /
+		    (c->trigger_mid_range - c->trigger_min_range) * 0.5f;
+	} else {
+		t = 0.5f - ((float)msg->touch.trigger - c->trigger_mid_range) /
+		    (c->trigger_max_range - c->trigger_mid_range) * 0.5f;
+	}
+	if (t != touch->trigger)
```

Comment:
Nitpick: the other values are assigned unconditionally. Maybe change this as well for consistency.

Reply:
Fixed